source/code-generators.ts

```diff
@@ -0,0 +1,101 @@
+import { dump, FAILSAFE_SCHEMA } from 'js-yaml';
+
+import { JSONObject, set } from './utils';
+
+function getDiff(locale: string, diffWith?: string) {
+  const diff: JSONObject = {};
+  const diffKeys = i18n.getAllKeysForLocale(diffWith);
+  i18n.getAllKeysForLocale(locale).forEach(key => {
+    if (diffKeys.includes(key)) {
+      set(diff, key, i18n.getTranslation(key));
+    }
+  });
+
+  return diff;
+}
+
+function getCachedFormatter(
+  type: 'json' | 'yml',
+  format: (translations: JSONObject) => string,
+) {
+  function cacheEntry(locale: string, namespace?: string, diffWith?: string) {
+    if (typeof namespace === 'string' && namespace) {
+      return {
+        key: `_${type}${namespace}`,
+        get: () =>
+          format({
+            _namespace: namespace,
+            ...((i18n.getTranslations(namespace, locale) as object) || {}),
+          }),
+      };
+    }
+
+    if (typeof diffWith === 'string' && diffWith) {
+      return {
+        key: `_${type}_diff_${diffWith}`,
+        get: () => format(getDiff(locale, diffWith)),
+      };
+    }
+
+    return {
+      key: `_${type}`,
+      get: () => format((i18n._translations[locale] as JSONObject) || {}),
+    };
+  }
+
+  return function cached(
+    locale: string,
+    localeCache: Record<string, string>,
+    namespace?: string,
+    diffWith?: string,
+  ) {
+    const { get, key } = cacheEntry(locale, namespace, diffWith);
+    if (!(key in localeCache)) {
+      localeCache[key] = get();
+    }
```

Comment:
Previously the cached objects were exposed through `i18n.getCache`, and as I checked, one of our projects relies on that (it clears it). Right now, this cache is not clearable, so it's a rather breaking change.

Is there any particular reason why you changed that?

Reply:
Previously the `getCachedFormatter` function was in the `server.ts` file and accessed the `cache` object directly to get the `localeCache` value. Since this function is now kept in a separate file it cannot access the `cache` object as previously, so instead it takes that as an argument.

But when we are [calling this function](https://github.com/vazco/meteor-universe-i18n/blob/cache-fix/source/server.ts#LL176C9-L176C80) we actually [use `cache`](https://github.com/vazco/meteor-universe-i18n/blob/cache-fix/source/server.ts#L167) exposed via the `i18n.getCache`. Which change breaks the cache-clearing logic?

Reply:
> Previously the `getCachedFormatter` function was in the `server.ts` file and accessed the `cache` object directly to get the `localeCache` value. Since this function is now kept in a separate file it cannot access the `cache` object as previously, so instead it takes that as an argument.

It can be accessed using `i18n.getCache()` now.

> But when we are [calling this function](https://github.com/vazco/meteor-universe-i18n/blob/cache-fix/source/server.ts#LL176C9-L176C80) we actually [use `cache`](https://github.com/vazco/meteor-universe-i18n/blob/cache-fix/source/server.ts#L167) exposed via the `i18n.getCache`. Which change breaks the cache-clearing logic?

The thing is, it's spread now, so writing to it won't change the `cache` object at all.

Reply:
Right, that makes sense - I wanted to black-box the `getCachedFormatter` function as much as possible since we already are accessing the cache before calling it, but it doesn't make sense to do that since getting cache from `i18n.getCache` is way simpler.